sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -846,27 +846,34 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
     return write_len;
 }
 
-ssize_t coap_opt_add_uquery(coap_pkt_t *pdu, const char *key, const char *val)
+ssize_t coap_opt_add_uquery(coap_pkt_t *pkt, const char *key, const char *val)
 {
+    return coap_opt_add_uquery2(pkt, key, strlen(key), val, val ? strlen(val) : 0);
+}
+
+ssize_t coap_opt_add_uquery2(coap_pkt_t *pkt, const char *key, size_t key_len,
+                             const char *val, size_t val_len)
+{
+    assert(pkt);
+    assert(key);
+    assert(key_len);
+
     char qs[NANOCOAP_QS_MAX];
-    size_t len = strlen(key);
-    size_t val_len = (val) ? (strlen(val) + 1) : 0;
+    size_t qs_len = key_len + ((val && val_len) ? 1 : 0) + val_len + 1;
```

Comment:
this looks overly complicated to me, I'd suggest:
```
size_t  qs_len = key_len + (val ? (val_len + 1) : 0);
```

the extra `+1` is not needed, see below

Reply:
the `assert` above helps here too, bc without the assert one could pass `val = NULL`  and `val_len = 42`, which might not be good

Reply:
yep that is simpler

Reply:
changed to the following to ignore `val_len` if `val` is `NULL`:
```C
size_t qs_len = key_len + ((val && val_len) ? (val_len + 1) : 0);
```

Reply:
I agree with this 100%! It is simpler more idiomatic to avoid the terminator in these calculations.